File: source/_posts/1.leetcode/数据结构/二叉树/遍历/94.二叉树的中序遍历.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=94 lang=java
 *
 * [94] 二叉树的中序遍历
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();
        while (root != null) {
            while (root != null) {
                stack.add(root);
                root = root.left;
            }
            stack.pop();
            res.add(root.val);
            System.out.println(root.val);
            root = root.right;
        }

    }

    public static void main(String[] args) {

    }

    List<Integer> res = new LinkedList<>();

    /**
     * 因为返回的是List, 拼接数组浪费，定义为全局list
     * 
     * @param root
     * @return
     */
    public List<Integer> inorderTraversal2(TreeNode root) {
        inorder(root);
        return res;

    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        res.add(root.val);
        inorder(root.right);
    }

}
// @lc code=end

File: source/_posts/Algorithm/1.leetcode/数据结构/二叉树/遍历/94.二叉树的中序遍历.py
#
# @lc app=leetcode.cn id=94 lang=python3
#
# [94] 二叉树的中序遍历
#

# @lc code=start
# Definition for a binary tree node.
from typing import List
# 如果一个树遍历的结果是有序数组，那么他也是一个二叉查找树(BST)


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        WHITE, GRAY = 0, 1
        res = []
        stack = [(WHITE, root)]
        while stack:
            color, node = stack.pop()
            if node is None:
                continue
            if color == WHITE:
                # 如果遇见白色，则继续深入
                stack.append((WHITE, node.right))
                # 中序遍历
                stack.append((GRAY, node))
                stack.append((WHITE, node.left))
            else:
                # 加入返回列表
                res.append(node.val)
                # print(node.val + ",")
        return res

    def inorderTraversal2(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        return self.inorderTraversal2(
            root.left) + [root.val] + self.inorderTraversal2(root.right)


# @lc code=end

test = Solution()

# test.inorderTraversal(root)
# print(res)